### PR TITLE
Add RDP (readout protection) support for STM32L4x flash

### PR DESCRIPTION
**Pull request by rlezuo1**
Add support for Flash readout protection on the STM32L4x series

### CODE REVIEW

**Review by de-nordic — DISMISSED** [review]
What I can see is that the provided functions are not used anywhere, there is no test for them there is no sample, and they have no external declarations provided.
What are they for?
If they are for users, then they should be implemented as `flash_ex_op`, as a vendor extension (include/zephyr/drivers/flash/stm32_flash_api_extensions.h).
Is this similar work as here https://github.com/zephyrproject-rtos/zephyr/pull/52980 and here https://github.com/zephyrproject-rtos/zephyr/pull/53441 ?

**Comment by erwango** [comment]
> I agree that samples and tests are missing (as for the existing code).

Tests are available in `tests/drivers/flash/stm32/src/main.c`

File: drivers/flash/Kconfig.stm32
Comment (erwango) [inline drivers/flash/Kconfig.stm32]: Please move `select  FLASH_HAS_EX_OP` to `config FLASH_STM32_READOUT_PROTECTION/FLASH_STM32_WRITE_PROTECT`, this should simplify things in the end.

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32l4x.c
Comment (de-nordic) [reply drivers/flash/flash_stm32l4x.c:297]: > `|| 1` looks suspect

That is in case the first part of equation is wrong ;)

File: tests/drivers/flash/stm32/testcase.yaml
Comment (rlezuo1) [reply tests/drivers/flash/stm32/testcase.yaml:17]: done

File: drivers/flash/Kconfig.stm32
Comment (rlezuo1) [reply drivers/flash/Kconfig.stm32]: done

**Review by erwango — DISMISSED** [review]

File: drivers/flash/flash_stm32l4x.c
Comment (de-nordic) [inline drivers/flash/flash_stm32l4x.c:318]: I wonder whether pairing _ASSERT and `return -ENOTSUP` makes sense.
Shouldn't here be rather `ERR_LOG` ?
Assuming that you have __ASSERT enabled for tests only, then why not just instead turn it off and fail the test if `-ENOTSUP` gets returned? You just want to catch problem and never get into it again.
In case when you have __ASSERT enabled in field, by accident, this halts your board.
On the other hand, if you have LOG_ERR here, in case when this fails in field you will have error log but the app may keep executing, so you can pull the logs and figure where the problem is.

File: drivers/flash/flash_stm32l4x.c
Comment (de-nordic) [inline drivers/flash/flash_stm32l4x.c:295]: What magic is happening here that the ifdef if making this comment conditional?

Comment (rlezuo1) [reply drivers/flash/flash_stm32l4x.c:295]: That would the place to implement WRITE_PROTECT on the L4 parts, it differs from the F4 parts significantly and this comments describes the things that would need to be done.

File: drivers/flash/flash_stm32l4x.c
Comment (rlezuo1) [reply drivers/flash/flash_stm32l4x.c:318]: that code is in sync with drivers/flash/flash_stm32f4x.c module some tricky to spot differences in register names OPTR vs OPTCR. 
If following your suggesting (I see your point here) I would unify those functions and place them into flash_stm32_ex_op_common.c. Would that make sense?
OTOH we could merge as-is and address this in the future.

File: drivers/flash/flash_stm32l4x.c
Comment (semihalf-duda-patryk) [inline drivers/flash/flash_stm32l4x.c:292]: Instead of implementing complex and error prone translation layer, you should add separate write protect extended operation for stm32 chips which have range based write protect.

File: drivers/flash/flash_stm32l4x.c
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32l4x.c:318]: > I would unify those functions and place them into flash_stm32_ex_op_common.c

`flash_stm32_ex_op_common.c` was created for keeping extended operations only. I don't think it's a good place to put here functions that touches the hardware.

File: drivers/flash/flash_stm32l4x.c
Comment (rlezuo1) [reply drivers/flash/flash_stm32l4x.c:292]: something like FLASH_STM32_EX_OP_SECTOR_WP_RANGED in include/zephyr/drivers/flash/stm32_flash_api_extensions.h?

File: drivers/flash/flash_stm32l4x.c
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32l4x.c:318]: Regarding `__ASSERT`. It doesn't use flash space when it's not enabled. Flash settings are not changed very often. One can enable asserts to check if there is an error.

File: drivers/flash/flash_stm32l4x.c
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32l4x.c:292]: Yes.

File: drivers/flash/flash_stm32l4x.c
Comment (rlezuo1) [reply drivers/flash/flash_stm32l4x.c]: I force pushed an update addressing this, I also think that LOG_ERR is better suited to __ASSERT

**Comment by rlezuo1** [comment]
@semihalf-duda-patryk @erwango friendly reminder

File: drivers/flash/flash_stm32f4x.c
Comment (erwango) [inline drivers/flash/flash_stm32f4x.c:296]: I won't block on this but ideally, all changes non strictly related to the new functionality and impacting existing code (ie which can cause a regression) should be made on a different commit. This should ease review and allow selected revert in case of regression.

File: tests/drivers/flash/stm32/testcase.yaml
Comment (erwango) [inline tests/drivers/flash/stm32/testcase.yaml:13]:
```suggestion
    filter: dt_compat_enabled("st,stm32f4-flash-controller") and
```

File: tests/drivers/flash/stm32/testcase.yaml
Comment (erwango) [inline tests/drivers/flash/stm32/testcase.yaml:21]:
```suggestion
    filter: dt_compat_enabled("st,stm32l4-flash-controller") and
```

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

File: drivers/flash/flash_stm32f4x.c
Comment (semihalf-duda-patryk) [inline drivers/flash/flash_stm32f4x.c:295]: I wonder if we really need `LOG_ERR` here. These messages consumes flash, but in well written application they will be never printed. How about changing it `LOG_WRN`? What do you think?

Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32f4x.c:295]: Okay.

**Review by duda-patryk — APPROVED** [review]

**Comment by rlezuo1** [comment]
@de-nordic friendly poke - with your approval the PR can be merged ...

**Review by de-nordic — APPROVED** [review]

**Comment by rlezuo1** [comment]
@erwango friendly poke: can you merge this PR?

**Comment by erwango** [comment]
> @erwango friendly poke: can you merge this PR?

No. We're in feature freeze for V3.5.0 stabilisation period. It will have to wait until merge window is open again.
See https://github.com/zephyrproject-rtos/zephyr/wiki/Release-Management

**Comment by rlezuo1** [comment]
@erwango full ack, sorry for the noise

**Pull request merged**
